### DISCUSSION TITLE
Add support for tablet relative mode

### DISCUSSION THREAD

**Discussion by CheesecakeCG** [YaLTeR/niri#711] [Ideas]
It looks like there was some support for this before but it was removed in https://github.com/YaLTeR/niri/issues/85 

Relative mode is pretty much a requirement for using drawing tablets for shooter games and some other genres and for me it makes digital sculpting much easier. When drawing or sculpting, it lets you use much more of the tablet's surface in a single stroke.

**Comment by YaLTeR** [comment]
Uhh? Is this something a compositor is even involved with? #85 is not really related.

Maybe you need https://opentabletdriver.net? If you don't map the tablet to any output in the niri config, it will map itself as OpenTabletDriver expects, so you should be able to use the OpenTabletDriver options.

**Reply by CheesecakeCG** [reply]
I think so, Hyprland and swaywm have config options for this.

I think I misread that issue when I skimmed over it, I'd vaguely remembered relative mode working (or more, didn't noticed it didn't default to absolute mode), but I'm probably misremembering things, my bad.

I've been using the digimend kernel drivers, I'll give OpenTabletDriver a try and see how it goes. 

Edit: OpenTabletDriver works with this tablet and switching to relative mode works, but there's no pressure sensitivity in relative mode and no mouse button support in "Artist Mode", and [their docs](https://opentabletdriver.net/Wiki/FAQ/Linux#artist-mode) say to switch to Artist Mode to get pressure sensitivity and to switch to Relative or Absolute mode to get mouse button input...

**Reply by YaLTeR** [reply]
Need to check what this sway option does in the code

**Reply by CheesecakeCG** [reply]
I did some more digging into this (this is now the only thing stopping me from switching to Niri full time). And I put together [a proof of concept](https://github.com/niri-wm/niri/compare/main...CheesecakeCG:niri:tablet-relative-mode). ~~I thought the code was broken, but it looks like half of my tablet died...~~ This works for well enough for me to stop yak-shaving and actually sculpt something though, minus the broken tablet.

From what I gather, a lot of the regular mouse handling code needs to be pulled out of `on_pointer_motion` in src/input/mod.rs to handle both relative mode tablets and mice, and/or smithay's `TabletToolHandle` also needs a `motion_relative` function like the pointer has?

Also the link to sway's implementation for reference: https://github.com/swaywm/sway/blob/0356a020c1c14b4f9e9c3583d6a87dd7764714a2/sway/input/cursor.c#L549

Edit: It was a firmware issue, not a hardware issue. Opening HuionTablet software in Windows fixed the issue I was having.

**Reply by billksun** [reply]
> If you don't map the tablet to any output in the niri config, it will map itself as OpenTabletDriver expects, so you should be able to use the OpenTabletDriver options.

What does this mean? If you don't map the tablet in niri config, doesn't the tablet get mapped to the union of all the outputs? It's still absolutely mapped and it will conflict with OTD's relative mode and cause the pointer to bounce around the screen like crazy.

@CheesecakeCG , is your PoC something that can be turned into a PR? I would love to not have to `sudo modprobe --remove wacom` and fire up OTD every time I need use relative mode.

**Reply by CheesecakeCG** [reply]
> @CheesecakeCG , is your PoC something that can be turned into a PR? I would love to not have to `sudo modprobe --remove wacom` and fire up OTD every time I need use relative mode.

It could be, but there's a lot of duplicate code between handling tablets & handling mice and the easiest way of adding relative mode would mean even more. I'm not 100% sure if I should try to refactor it or what approach I should take if I do.
